File: backend-gin/internal/models/examples/seed.go
package examples

import (
	"backend/internal/models/cruds"
	"backend/internal/models/orm"
	"backend/internal/models/schemas"
	"backend/internal/services/instances"
	"context"
	"errors"
	"fmt"
	"sync"

	"golang.org/x/sync/errgroup"
)

type RefMappings map[orm.Model]map[int]uint

func seedUsers(ctx context.Context, refs RefMappings, isVerbose bool) error {
	userRefs := make(map[int]uint)
	refs[orm.ModelUser] = userRefs
	cu := cruds.GetCRUDSUser()
	storage := instances.GetStorage()

	handleError := func(err error, isVerbose bool) error {
		if isVerbose {
			fmt.Println(err.Error())
		}
		return fmt.Errorf(
			"could not create user document: %w", err,
		)
	}

	var eg errgroup.Group
	var mu sync.Mutex
	for _, userEx := range Users {
		example := userEx // capture loop variable
		eg.Go(func() error {
			userIn := schemas.UserCreate{
				Name:     userEx.Name,
				Email:    userEx.Email,
				IsAdmin:  userEx.IsAdmin,
				Password: userEx.Password,
				ImageURL: userEx.ImageURL,
			}

			url, err := storage.UploadFile(ctx, userIn.ImageURL, "")
			if err != nil {
				return handleError(err, isVerbose)
			}
			userIn.ImageURL = url

			insertedId, err := cu.Create(ctx, userIn)
			if err != nil {
				return handleError(err, isVerbose)
			}

			mu.Lock()
			userRefs[example.Ref] = insertedId
			mu.Unlock()
			return nil
		})
	}

	return eg.Wait()
}

func seedPlaces(ctx context.Context, refs RefMappings, isVerbose bool) error {
	placeRefs := make(map[int]uint)
	refs[orm.ModelPlace] = placeRefs
	cp := cruds.GetCRUDSPlace()
	storage := instances.GetStorage()

	handleError := func(err error, isVerbose bool) error {
		if isVerbose {
			fmt.Println(err.Error())
		}
		return fmt.Errorf(
			"could not create place document: %w", err,
		)
	}

	var eg errgroup.Group
	var mu sync.Mutex
	for _, placeEx := range Places {
		example := placeEx // capture loop variable
		eg.Go(func() error {
			placeIn := schemas.PlaceCreate{
				Title:       placeEx.Title,
				Description: placeEx.Description,
				Address:     placeEx.Address,
				Location:    placeEx.Location,
				ImageURL:    placeEx.ImageURL,
			}

			url, err := storage.UploadFile(ctx, placeIn.ImageURL, "")
			if err != nil {
				return handleError(err, isVerbose)
			}
			placeIn.ImageURL = url

			creatorId, found := refs[orm.ModelUser][placeEx.CreatorRef]
			if !found {
				message := fmt.Sprintf(
					"examples corrrupted, could not find user with ref %d while creating place with ref %d",
					placeEx.CreatorRef,
					placeEx.Ref,
				)
				if isVerbose {
					fmt.Println(message)
				}
				return errors.New(message)
			}
			placeIn.CreatorID = creatorId

			insertedId, err := cp.Seed(ctx, placeIn, placeEx.Embedding)
			if err != nil {
				return handleError(err, isVerbose)
			}

			mu.Lock()
			placeRefs[example.Ref] = insertedId
			mu.Unlock()
			return nil
		})
	}

	return eg.Wait()
}

func SeedDb(verbose ...bool) error {
	ctx := context.Background()
	refs := make(RefMappings)

	isVerbose := false
	if len(verbose) > 0 {
		isVerbose = verbose[0]
	}

	if err := seedUsers(ctx, refs, isVerbose); err != nil {
		if isVerbose {
			fmt.Println(err.Error())
		}
		return err
	} else if isVerbose {
		fmt.Println("✅ Collection User seeded!")
	}

	if err := seedPlaces(ctx, refs, isVerbose); err != nil {
		if isVerbose {
			fmt.Println(err.Error())
		}
		return err
	} else if isVerbose {
		fmt.Println("✅ Collection Place seeded!")
	}

	if isVerbose {
		fmt.Println("✅ Finished. You may exit")
	}

	return nil
}

func DumpDb(verbose ...bool) error {
	isVerbose := false
	if len(verbose) > 0 {
		isVerbose = verbose[0]
	}

	ctx := context.Background()
	pgClient := instances.GetPgClient()

	for _, table := range orm.AllTables {
		err := pgClient.ResetTable(ctx, string(table))
		if err != nil {
			message := fmt.Sprintf("could not reset table %s", table)
			if isVerbose {
				fmt.Println(message)
			}
		} else if isVerbose {
			fmt.Printf("✅ Table %s cleared!\n", table)
		}
	}

	rc := instances.GetRedisClient()
	rc.FlushAll(ctx)

	if isVerbose {
		fmt.Println("✅ Cache DB flushed")
		fmt.Println("✅ Finished. You may exit")
	}

	return nil
}

File: backend-axum/src/models/examples/seed.rs
use std::{collections::HashMap, sync::Arc};

use tokio::task::JoinSet;

use crate::{
    config,
    lib_::{seaorm_::Create, types_::FileToUpload, utils},
    models::{
        cruds::{CrudsPlace, CrudsUser},
        examples::{place::get_place_seeds, user::get_user_seeds},
        orm::{PLACE_MODEL, USER_MODEL, get_tables},
        schemas::{PlacePost, UserPost},
    },
    services::instances::AppState,
};

type RecordMap = HashMap<u32, u32>;
type RecordsMap = HashMap<String, RecordMap>;

async fn seed_users(state: Arc<AppState>, refs: &mut RecordsMap) -> Result<(), String> {
    let cruds = CrudsUser::new(state);
    let mut join_set = JoinSet::new();

    for user in get_user_seeds() {
        let cruds = cruds.clone();
        join_set.spawn(async move {
            // Step 1: hash password
            let hashed = utils::hash_input(&user.password, config::ENV.default_hash_salt as u32)
                .expect("failed to hash a password");

            // Step 2: create the post form
            let post_form = UserPost {
                name: user.name.clone(),
                email: user.email.clone(),
                is_admin: user.is_admin,
                password: hashed,
                image: Some(
                    FileToUpload::from_path(&user.image_url)
                        .expect("failed to load an image file for user record"),
                ),
            };

            // Step 3: convert the Post to Create (handle image upload)
            let data = cruds
                .post_to_create(post_form)
                .await
                .expect("failed to build creation form for user record");

            // Step 4: create the record
            let id = cruds
                .create(data)
                .await
                .expect("failed to create a user record");
            (user.ref_, id)
        });
    }

    let mut map = HashMap::new();
    while let Some(result) = join_set.join_next().await {
        match result {
            Ok((ref_, id)) => {
                map.insert(ref_, id);
            }
            Err(e) => return Err(format!("Task panicked: {:?}", e)),
        }
    }
    refs.insert(USER_MODEL.to_string(), map);
    Ok(())
}

async fn seed_places(state: Arc<AppState>, refs: &mut RecordsMap) -> Result<(), String> {
    let cruds = CrudsPlace::new(state);
    let mut join_set = JoinSet::new();
    let users_map = Arc::new(
        refs.get(USER_MODEL)
            .expect("failed to extract the users id mapping")
            .clone(),
    );

    for place in get_place_seeds() {
        let cruds = cruds.clone();
        let users_map = users_map.clone();
        join_set.spawn(async move {
            // Step 1: get the corresponding creator_id
            let creator_id = users_map
                .get(&place.creator_ref)
                .expect("failed to extract a user from a place creator_ref");

            // Step 2: create the post form
            let post_form = PlacePost {
                creator_id: *creator_id,
                title: place.title.clone(),
                description: place.description.clone(),
                address: place.address.clone(),
                lat: place.location.lat,
                lng: place.location.lng,
                image: Some(
                    FileToUpload::from_path(&place.image_url)
                        .expect("failed to load an image file for place record"),
                ),
            };

            // Step 3: convert the Post to Create (handle image upload)
            let data = cruds
                .post_to_create(post_form)
                .await
                .expect("failed to build creation form for place record");

            // Step 4: create the record
            let id = cruds
                .seed(data, place.embedding)
                .await
                .expect("failed to create a place record");

            (place.ref_, id)
        });
    }

    let mut map = HashMap::new();
    while let Some(result) = join_set.join_next().await {
        match result {
            Ok((ref_, id)) => {
                map.insert(ref_, id);
            }
            Err(e) => return Err(format!("Task panicked: {:?}", e)),
        }
    }
    refs.insert(PLACE_MODEL.to_string(), map);
    Ok(())
}

pub async fn seed_db(verbose: bool) {
    let app_state = Arc::new(AppState::new().await);
    let mut refs: RecordsMap = HashMap::new();

    // Users
    seed_users(app_state.clone(), &mut refs)
        .await
        .expect("failed at seeding user records");
    if verbose {
        println!("✅ Users data seeded!");
    }

    // Places
    seed_places(app_state.clone(), &mut refs)
        .await
        .expect("failed at seeding place records");
    if verbose {
        println!("✅ Places data seeded!");
    }

    if verbose {
        println!("✅ Finished. You may exit");
    }
}

pub async fn dumb_db(verbose: bool) {
    let app_state = AppState::new().await;
    for tablename in get_tables() {
        let result = app_state.pg.reset_table(tablename).await;
        match result {
            Ok(_) => {
                if verbose {
                    println!("✅ Table {} cleared!", tablename);
                }
            }
            Err(err) => {
                if verbose {
                    println!("Could not reset table {}", tablename);
                    println!("{:?}", err);
                }
                return;
            }
        }

        if result.is_err() && verbose {}
    }

    let result = app_state.redis.flush_all().await;
    match result {
        Ok(_) => {
            if verbose {
                println!("✅ Cache DB flushed");
                println!("✅ Finished. You may exit");
            }
        }
        Err(err) => {
            if verbose {
                println!("Failed to flush redis db");
                println!("{:?}", err);
            }
        }
    }
}
